File: 01-coding/golang/hello_sre.go
package main

/*
Hello SRE - Go Edition

This demonstrates basic Go programming for SREs:
- Structs and methods
- Error handling
- JSON encoding
- Concurrent operations
*/

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"time"
)

// SystemInfo represents system information
type SystemInfo struct {
	Hostname     string    `json:"hostname"`
	OS           string    `json:"os"`
	Architecture string    `json:"architecture"`
	NumCPU       int       `json:"num_cpu"`
	GoVersion    string    `json:"go_version"`
	Timestamp    time.Time `json:"timestamp"`
}

// GetSystemInfo gathers system information
func GetSystemInfo() (*SystemInfo, error) {
	hostname, err := os.Hostname()
	if err != nil {
		return nil, fmt.Errorf("failed to get hostname: %w", err)
	}

	info := &SystemInfo{
		Hostname:     hostname,
		OS:           runtime.GOOS,
		Architecture: runtime.GOARCH,
		NumCPU:       runtime.NumCPU(),
		GoVersion:    runtime.Version(),
		Timestamp:    time.Now(),
	}

	return info, nil
}

// PrintSystemInfo displays system information in a formatted way
func PrintSystemInfo(info *SystemInfo) {
	fmt.Println("\n" + "="*50)
	fmt.Println("🚀 Hello from SRE Training!")
	fmt.Println("="*50)
	fmt.Println("\n📊 System Information:")
	fmt.Printf("  Hostname: %s\n", info.Hostname)
	fmt.Printf("  OS: %s\n", info.OS)
	fmt.Printf("  Architecture: %s\n", info.Architecture)
	fmt.Printf("  CPU Cores: %d\n", info.NumCPU)
	fmt.Printf("  Go Version: %s\n", info.GoVersion)
	fmt.Printf("  Timestamp: %s\n", info.Timestamp.Format(time.RFC3339))
	fmt.Println("\n" + "="*50)
}

// ExportJSON exports system info as JSON
func ExportJSON(info *SystemInfo) (string, error) {
	data, err := json.MarshalIndent(info, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal JSON: %w", err)
	}
	return string(data), nil
}

func main() {
	// Gather system information
	info, err := GetSystemInfo()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	// Print formatted output
	PrintSystemInfo(info)

	// Export as JSON
	jsonOutput, err := ExportJSON(info)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error exporting JSON: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("📄 JSON Export:")
	fmt.Println(jsonOutput)
}

File: 01-coding/python/hello_sre.py
#!/usr/bin/env python3
"""
Hello SRE - Python Edition

This is a simple introduction to Python scripting for SREs.
Demonstrates basic concepts: logging, error handling, and system information gathering.
"""

import logging
import platform
import sys
from datetime import datetime

# Configure logging - essential for operational scripts
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


def get_system_info():
    """
    Gather system information - a common SRE task.
    
    Returns:
        dict: System information
    """
    try:
        info = {
            'hostname': platform.node(),
            'os': platform.system(),
            'os_version': platform.release(),
            'architecture': platform.machine(),
            'python_version': platform.python_version(),
            'timestamp': datetime.now().isoformat()
        }
        return info
    except Exception as e:
        logger.error(f"Error gathering system info: {e}")
        raise


def main():
    """Main function with proper error handling."""
    try:
        logger.info("Starting SRE Hello World script")
        
        # Gather system information
        sys_info = get_system_info()
        
        # Display results
        print("\n" + "="*50)
        print("🚀 Hello from SRE Training!")
        print("="*50)
        print(f"\n📊 System Information:")
        for key, value in sys_info.items():
            print(f"  {key}: {value}")
        print("\n" + "="*50)
        
        logger.info("Script completed successfully")
        return 0
        
    except Exception as e:
        logger.error(f"Script failed: {e}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
